### ISSUE TITLE
Type of generic function as unknown when using tsx component

### ISSUE THREAD

**Issue by s-montigny-desautels** [vuejs/language-tools#3309]
### Vue version

3.3.4

### Description
This issue appears when importing a `.tsx` component in a `.vue` component. If the props contain a function with a generic argument, the argument is of type `unknown`.

### Link to minimal reproduction

https://github.com/s-montigny-desautels/vue3.3-issues

### Steps to reproduce

```bash
$ pnpm install
```

Look for the error in file `BugTSXFunctionUnknown.vue`

### What is expected?

The type of the function to respect the generic

### What is actually happening?

The type of the function argument is of type `unknown`

### Any additional comments?
This issue is only present in a `.vue` file. When using `.tsx` file, everything work as expected.

**Comment by rijenkii** [comment]
Smaller reproduction of the seemingly same problem:
```vue
<script setup lang="ts">
import { ref } from 'vue';

const val = ref(0);
const test = <K,>(_: { modelValue: K, "onUpdate:modelValue"?: (value: K) => any }) => "";
</script>

<template>
  <test :model-value="123" @update:model-value="val = $event" />
</template>
```
[playground](https://play.vuejs.org/#eNp9UctOwzAQ/JWVhdRWKi3QW0jKSz1AJUC8TpZQlG5DimNbfoSiKv/O2lEfB9RLZM+MZ2c2G3aj9ajxyBKW2sJU2oFF5zWIXJYZZ85yNuWyqrUyDjZgcAktLI2qoUfPepdcclkoaR00uYAsCPpnA4I70CF9Mkjnw2n/MyGDWi1QfOTCYwLzIXCm5Lte5A6TPcPZVQJ98ouiAWRTyOUvtPHEGWdkn467uBSOLg5rLciEbgB0o6Gd32l0oSLnFxPO4NofzNpxXfITbFA6Eo3JJR0fWLIh7YH6LKtytLJK0rI2YRBnhap1JdA8aVdRX86oYmAClwuhfh4i5ozH4RYvvrD4/gdf2XXAOHs2aNE0tIYd53JTImUL9Oz1Edd03pFUxgtSHyFf0CrhQ8ZOduvlgmIf6GLa+/ibK1m+2dnaobTbUiFoULZRTxvzeHek+j7uZDSJ77hsWfsHTlXLcw==)

Error in `@update:model-value`: `Type 'unknown' is not assignable to type 'number'. ts(2322)`.
When hovering over `:model-value` and `$event` their types are shown as `unknown`.

So, it seems that the problem is not with JSX components, but with all generic functional components.

**Comment by rijenkii** [comment]
However, bizzarly, type checking in `.tsx` files works without a hitch:
```tsx
import { ref } from "vue";

const val = ref(0);
const Test = <K,>(_: { modelValue: K; "onUpdate:modelValue"?: (value: K) => any }) =>
  null;

export default () => (
  <Test modelValue={val.value} onUpdate:modelValue={(event) => (val.value = event)} />
);
```

**Comment by rijenkii** [comment]
Current workaround is: Make a factory:

```vue
<script setup lang="ts">
import { ref } from 'vue';

const val = ref(0);
const test = <K,>(_: { modelValue: K, "onUpdate:modelValue"?: (value: K) => any }) => "";

const testFactory = <K,>() => test<K>;

const testNumber = testFactory<number>();
</script>

<template>
  <testNumber :model-value="123" @update:model-value="val = $event" />
</template>
```

[playground](https://play.vuejs.org/#eNp9UttqAjEQ/ZUhFKpgta1v29XesNAKtvT2FCjbdbRrs0nIxSqy/95JFnUpxZclOefMzJmz2bBrrbtLjyxhqc1NoR1YdF6DyOR8wJmznA25LEqtjIMNGJxBBTOjSjimsuMLLrnMlbQOlpmAQRC0TtsE16BD+gwgHXeGrY+EGpRqiuI9Ex4TGHeAMyXf9DRzmOwZzi4TaFG/KGrDYAiZXEMVT5xx1pgaBtxluVNmvZ0TZQFPx8M/yokvP9GQsFGWyghSHYnTXp0C7UwXh6UW5I1uAHTbNajNnkSLlNLZeZ8zuPKNRXZcHcsRLlE6EvWoV9prNGYdCpkczop5d2GVpD+xCeM4y1WpC4HmUbuCNuCM8gtM4DIh1M9DxJzx2Nni+Rfm3//gC7sKGGdPBi2aJWW841xm5kjeAj16meCKzjuSlvGC1AfIZ7RK+OCxlt14OSXbDV10ex/fUCHnr3a0cijtdqlgNCirqKfEPN4eWH1vt9/txzouK1b9Ag7e7DE=)

**Comment by s-montigny-desautels** [comment]
Fixed with Vue 3.4.20
